### DISCUSSION TITLE
Failure on upgrade

### DISCUSSION THREAD

**Discussion by airigami** [mcguirepr89/BirdNET-Pi#275] [General]
I am really loving playing with BirdNET-Pi. Thank you mcquirepr89 and everyone else that's contributed. I apologize for the newb question, but I didn't find an answer looking around. I have a very limited understanding of how GitHub works. 

After a week or so of watching things run, I noticed the Apprise notifications were added and wanted to play. I followed the update instructions in the readme.  When I do the git pull, I get:

error: Pulling is not possible because you have unmerged files.
hint: Fix them up in the work tree, and then use 'git add/rm <file>'
hint: as appropriate to mark resolution and make a commit.
fatal: Exiting because of an unresolved conflict.

I'm lost as to what I need to do. I haven't changed anything other than the config settings, so I shouldn't have anything odd. I can always backup my database, reinstall and restore the database, but I'd rather learn what's going on so I can fix stuff like this in the future.

Then I'll see if I can figure out how to set up notifications.

Thanks.

**Comment by mcguirepr89** [comment]
Hello, @airigami! Thanks for the kind words about the project. I think the mistake is mine for putting that section in the ReadMe.

Try upgrading with "Tools" > "System Controls" > "Update" and if you'll paste the output it posts here, we can take a look through and see if whether that upgrade method works without those git issues.

To elucidate the issue a bit more, there are customizations that happen in the scripts (the shebangs for the python virtual environment are added to the tops of the python scripts). `git` was complaining because you wanted to `pull` changes (because that's what the ReadMe instructs!! silly me) but haven't told `git` what you want it to do with your customizations. 

The answer is the "Tools" > "System Contols" > "Update" tools which calls the `update_birdnet.sh` script. This script aims to handle the git stuff by first
1. `git stash` which tells git you want to keep the custom shebangs for later
2. `git pull` which fetches and merges the new changes
3. `git stash pop` which puts the customizations back in place where it found them, hoping that the updated bits were not among those places customized.

I hope that helps you understand a bit behind what failed in your steps and why I'm changing the ReadMe now :)

Please let me know how the update goes from the web interface.

My best regards,
Patrick

**Reply by airigami** [reply]
Thanks for the quick reply. Here's what the update looks like, followed by a section of the log:

Saved working directory and index state WIP on main: 62ca6fc Update README.md
hint: Pulling without specifying how to reconcile divergent branches is
hint: discouraged. You can squelch this message by running one of the following
hint: commands sometime before your next pull:
hint: 
hint:   git config pull.rebase false  # merge (the default strategy)
hint:   git config pull.rebase true   # rebase
hint:   git config pull.ff only       # fast-forward only
hint: 
hint: You can replace "git config" with "git config --global" to set a default
hint: preference for all repositories. You can also pass --rebase, --no-rebase,
hint: or --ff-only on the command line to override the configured default per
hint: invocation.
Already up to date.
On branch main
Your branch is up to date with 'origin/main'.

Changes not staged for commit:
  (use "git add ..." to update what will be committed)
  (use "git restore ..." to discard changes in working directory)
	modified:   analyze.py
	modified:   daily_plot.py
	modified:   plotly_streamlit.py
	modified:   server.py

Untracked files:
  (use "git add ..." to include in what will be committed)
	../BirdDB.txt
	../HUMAN.txt
	../analyzing_now.txt
	../apprise.txt
	../birdnet/
	../exclude_species_list.txt
	../firstrun.ini
	../include_species_list.txt
	birds.db
	exclude_species_list.txt
	include_species_list.txt
	labels.txt
	lastrun.txt
	thisrun.txt
	../templates/avahi-alias@.service
	../templates/birdnet_analysis.service
	../templates/birdnet_log.service
	../templates/birdnet_recording.service
	../templates/birdnet_server.service
	../templates/birdnet_stats.service
	../templates/chart_viewer.service
	../templates/custom_recording.service
	../templates/extraction.service
	../templates/livestream.service
	../templates/pushed_notifications.service
	../templates/spectrogram_viewer.service
	../templates/web_terminal.service

no changes added to commit (use "git add" and/or "git commit -a")
Dropped refs/stash@{0} (d8959f41b18184ee5cec8b9d704eedcf14d518a1)

---

14:36:44 server.py[434804]:   File "/usr/local/bin/server.py", line 52, in <module>
14:36:44 server.py[434804]:     priv_thresh = float("." + str(str(str([i for i in this_run if i.startswith('PRIVACY_THRESHOLD')]).split('=')[1]).split('\\')[0]))/10
14:36:44 server.py[434804]: IndexError: list index out of range
14:36:44 systemd[1]: birdnet_server.service: Main process exited, code=exited, status=1/FAILURE
14:36:44 systemd[1]: birdnet_server.service: Failed with result 'exit-code'.
14:36:44 systemd[1]: birdnet_server.service: Consumed 4.092s CPU time.
14:36:54 systemd[1]: birdnet_server.service: Scheduled restart job, restart counter is at 12.
14:36:54 systemd[1]: Stopped BirdNET Analysis Server.
14:36:54 systemd[1]: birdnet_server.service: Consumed 4.092s CPU time.
14:36:54 systemd[1]: Started BirdNET Analysis Server.
14:36:58 server.py[435088]: Traceback (most recent call last):
14:36:58 server.py[435088]:   File "/usr/local/bin/server.py", line 52, in <module>
14:36:58 server.py[435088]:     priv_thresh = float("." + str(str(str([i for i in this_run if i.startswith('PRIVACY_THRESHOLD')]).split('=')[1]).split('\\')[0]))/10
14:36:58 server.py[435088]: IndexError: list index out of range
14:36:58 systemd[1]: birdnet_server.service: Main process exited, code=exited, status=1/FAILURE
14:36:58 systemd[1]: birdnet_server.service: Failed with result 'exit-code'.
14:36:58 systemd[1]: birdnet_server.service: Consumed 4.325s CPU time.
14:37:00 systemd[1]: Stopped BirdNET Analysis Server.
14:37:00 systemd[1]: birdnet_server.service: Consumed 4.325s CPU time.
14:37:00 systemd[1]: Started BirdNET Analysis Server.
14:37:04 server.py[435232]: Traceback (most recent call last):
14:37:04 server.py[435232]:   File "/usr/local/bin/server.py", line 52, in <module>
14:37:04 server.py[435232]:     priv_thresh = float("." + str(str(str([i for i in this_run if i.startswith('PRIVACY_THRESHOLD')]).split('=')[1]).split('\\')[0]))/10
14:37:04 server.py[435232]: IndexError: list index out of range
14:37:04 systemd[1]: birdnet_server.service: Main process exited, code=exited, status=1/FAILURE
14:37:04 systemd[1]: birdnet_server.service: Failed with result 'exit-code'.
14:37:04 systemd[1]: birdnet_server.service: Consumed 4.106s CPU time.
14:37:14 systemd[1]: birdnet_server.service: Scheduled restart job, restart counter is at 13.
14:37:14 systemd[1]: Stopped BirdNET Analysis Server.
14:37:14 systemd[1]: birdnet_server.service: Consumed 4.106s CPU time.
14:37:14 systemd[1]: Started BirdNET Analysis Server.
14:37:18 server.py[435528]: Traceback (most recent call last):
14:37:18 server.py[435528]:   File "/usr/local/bin/server.py", line 52, in <module>
14:37:18 server.py[435528]:     priv_thresh = float("." + str(str(str([i for i in this_run if i.startswith('PRIVACY_THRESHOLD')]).split('=')[1]).split('\\')[0]))/10
14:37:18 server.py[435528]: IndexError: list index out of range
14:37:18 systemd[1]: birdnet_server.service: Main process exited, code=exited, status=1/FAILURE
14:37:18 systemd[1]: birdnet_server.service: Failed with result 'exit-code'.
14:37:18 systemd[1]: birdnet_server.service: Consumed 4.286s CPU time.
14:37:20 systemd[1]: Stopped BirdNET Analysis Server.
14:37:20 systemd[1]: birdnet_server.service: Consumed 4.286s CPU time.

**Reply by mcguirepr89** [reply]
It may be required to restart all services after this update. I don't see anything in the output you provided that show errors in the upgrade, but the `list index out of range` makes it seem that there is at least something not right with your `server.py`.

Could you please go to "Tools" > "Services" > "Restart ALL Services" at the bottom and let me know after it restarts all services whether that clears up the issue with the `list index out of range`? Note that restarting all services takes anywhere from 30 - 90 seconds depending on how badly the socket wants to stay bound to the port.

**Reply by airigami** [reply]
I restarted all services and went to get a cup of tea, so plenty of time. Still seeing the failure message.

15:29:39 server.py[505094]: Traceback (most recent call last):
15:29:39 server.py[505094]:   File "/usr/local/bin/server.py", line 52, in <module>
15:29:39 server.py[505094]:     priv_thresh = float("." + str(str(str([i for i in this_run if i.startswith('PRIVACY_THRESHOLD')]).split('=')[1]).split('\\')[0]))/10
15:29:39 server.py[505094]: IndexError: list index out of range
15:29:40 systemd[1]: birdnet_server.service: Main process exited, code=exited, status=1/FAILURE
15:29:40 systemd[1]: birdnet_server.service: Failed with result 'exit-code'.
15:29:40 systemd[1]: birdnet_server.service: Consumed 4.500s CPU time.
15:29:50 systemd[1]: birdnet_server.service: Scheduled restart job, restart counter is at 1.
15:29:50 systemd[1]: Stopped BirdNET Analysis Server.
15:29:50 systemd[1]: birdnet_server.service: Consumed 4.500s CPU time.
15:29:50 systemd[1]: Started BirdNET Analysis Server.
15:29:51 systemd[1]: Stopping BirdNET Analysis Server...
15:29:51 systemd[1]: birdnet_server.service: Succeeded.
15:29:51 systemd[1]: Stopped BirdNET Analysis Server.
15:29:51 systemd[1]: Started BirdNET Analysis Server.
15:29:54 server.py[505432]: Traceback (most recent call last):
15:29:54 server.py[505432]:   File "/usr/local/bin/server.py", line 52, in <module>
15:29:54 server.py[505432]:     priv_thresh = float("." + str(str(str([i for i in this_run if i.startswith('PRIVACY_THRESHOLD')]).split('=')[1]).split('\\')[0]))/10
15:29:54 server.py[505432]: IndexError: list index out of range
15:29:55 systemd[1]: birdnet_server.service: Main process exited, code=exited, status=1/FAILURE
15:29:55 systemd[1]: birdnet_server.service: Failed with result 'exit-code'.
15:29:55 systemd[1]: birdnet_server.service: Consumed 4.153s CPU time.
15:29:56 systemd[1]: Stopped BirdNET Analysis Server.
15:29:56 systemd[1]: birdnet_server.service: Consumed 4.153s CPU time.
15:29:56 systemd[1]: Started BirdNET Analysis Server.
15:29:59 server.py[505568]: Traceback (most recent call last):
15:29:59 server.py[505568]:   File "/usr/local/bin/server.py", line 52, in <module>
15:29:59 server.py[505568]:     priv_thresh = float("." + str(str(str([i for i in this_run if i.startswith('PRIVACY_THRESHOLD')]).split('=')[1]).split('\\')[0]))/10
15:29:59 server.py[505568]: IndexError: list index out of range
15:30:00 systemd[1]: birdnet_server.service: Main process exited, code=exited, status=1/FAILURE
15:30:00 systemd[1]: birdnet_server.service: Failed with result 'exit-code'.
15:30:00 systemd[1]: birdnet_server.service: Consumed 4.124s CPU time.
15:30:10 systemd[1]: birdnet_server.service: Scheduled restart job, restart counter is at 1.
15:30:10 systemd[1]: Stopped BirdNET Analysis Server.
15:30:10 systemd[1]: birdnet_server.service: Consumed 4.124s CPU time.
15:30:10 systemd[1]: Started BirdNET Analysis Server.
15:30:11 systemd[1]: Stopping BirdNET Analysis Server...
15:30:11 systemd[1]: birdnet_server.service: Succeeded.
15:30:11 systemd[1]: Stopped BirdNET Analysis Server.
15:30:11 systemd[1]: birdnet_server.service: Consumed 1.200s CPU time.
15:30:11 systemd[1]: Started BirdNET Analysis Server.
15:30:14 server.py[505897]: Traceback (most recent call last):
15:30:14 server.py[505897]:   File "/usr/local/bin/server.py", line 52, in <module>
15:30:14 server.py[505897]:     priv_thresh = float("." + str(str(str([i for i in this_run if i.startswith('PRIVACY_THRESHOLD')]).split('=')[1]).split('\\')[0]))/10
15:30:14 server.py[505897]: IndexError: list index out of range
15:30:15 systemd[1]: birdnet_server.service: Main process exited, code=exited, status=1/FAILURE
15:30:15 systemd[1]: birdnet_server.service: Failed with result 'exit-code'.
15:30:15 systemd[1]: birdnet_server.service: Consumed 4.076s CPU time.
15:30:16 systemd[1]: Stopped BirdNET Analysis Server.
15:30:16 systemd[1]: birdnet_server.service: Consumed 4.076s CPU time.
15:30:16 systemd[1]: Started BirdNET Analysis Server.

**Reply by mcguirepr89** [reply]
Thanks for giving it a good try. I'm pretty stumped, but this is what I would try if I were you.

In the terminal:

```
git checkout -f
git pull
update_birdnet.sh && restart_services & journalctl -fu birdnet_server -ubirdnet_analysis
```

That should get your system working correctly. The last line does an update and service restart but watches the logs as that takes place. You'll need to `Ctrl + C` to quit after a while (hopefully once you start seeing normal log behavior).

If the steps above don't work, the second thing I would try would be to enable debugging in the `birdnet_analysis.sh` script. Removing the `#` comment of line 3 of that file should enable debugging in the log and help us troubleshoot further.

Sorry about this unexpected trouble!

**Reply by airigami** [reply]
I appreciate the time you've given this. Of course, now I've lost my connection to it. Not sure why. I'm doing this remotely, so I'll poke it again when I get home. If all else fails, I can reinstall. Any suggestions on the best way to do that? I figure I can backup the database and make a copy of the conf file. Anything else I need to save? I think the only thing I even really need to save from the conf is GPS location, which I can find again, and my bird weather login. Then can I just wipe that directory and rerun the initial install script?

(If you want to figure out what went wrong, I'm happy to keep trying things with you. I just don't know how much longer that makes sense for you.)

**Reply by mcguirepr89** [reply]
You're welcome for the support, I'm hopeful we can get it upgraded. And I'm certainly interested in finding out what has actually failed. It has something to do with `server.py`'s reading the `PRIVACY_THRESHOLD` variable.

When you do get reconnected and have time, let me know if you see anything funky in "Tools" > "Settings" > "Advanced Settings" (the funk factor in this case being an error, not something awesome like Funkadelic or Zapp). 

Also when able, if you could send me the output of:
```
grep -e '^PRIVACY' ~/BirdNET-Pi/birdnet.conf
```
that could help shed some light on the issue.

Lastly, if there is nothing strange going on with those things (all related to the update not populating your `birdnet.conf`'s `PRIVACY_THRESHOLD` variable correctly), I'll ask that you try this command:
```
awk '!/#/ && !/^$/ `{print}'` /etc/birdnet/birdnet.conf > ~/BirdNET-Pi/thisrun.txt && sudo systemctl restart_services.sh & birdnet_log.sh
```
and see whether _that_ works. That command just manually updates a file that gets updated each `birdnet_analysis.service` cycle, speculating that maybe it hadn't been updated yet?

The short re-installation guide can just be this one-liner if you would rather just change the lat and longitude after installation and re-add your BirdWeather ID:
```
cd && uninstall.sh && rm -drfv BirdNET* && curl -s https://raw.githubusercontent.com/mcguirepr89/BirdNET-Pi/main/newinstaller.sh | bash
```
Though, to answer your question about the best way to uninstall and reinstall, you'll want to make a backup of your database `~/BirdNET-Pi/scripts/birds.db`, run `uninstall.sh`, then `cd && rm -drfv BirdNET*`, then you can rerun the curl one-line installer and then move your database back into place.

**Comment by airigami** [comment]
The crash that took the system offline yesterday was due to filesystem corruption. An fsck repaired enough to make it mountable so I could grab stuff, but it still wasn’t booting up properly. And for the sake of future stability, it made more sense to reinstall. I reformatted the card and started from scratch. Put my old data back and had it working for a little while, still able to see all of my old charts. A few hours later, it’s offline again. I haven’t had a chance to look further, but since it was running perfectly and now won’t boot, my guess is that the root of the problem all along was a failing SD card. Or at least, a card that has failed since we tried making it all work again. So, tomorrow, or later tonight, I’ll start over with a new card and see where that gets me. I can temporarily grab another pi to use to see if that makes a difference too in case the pi I’m using has issues, but a bad SD card seems so much more likely.

Larry Moss, Creative Director
Airigami.com
510 State St, Suite 7
Rochester, NY 14608
+1.585.359.8695
Pronouns: he/him

**Reply by mcguirepr89** [reply]
:thinking: interesting observations. Along those lines, if the database or files themselves were a source of corruption, that may also explain how the crash appeared on the new system not long after restoring those files to the SD card.

Please keep me posted on how this is working out for you, as I want to know with more certainty whether we're dealing with a software bug or a hardware snafu.

Please accept my best regards,
Patrick

**Reply by ehpersonal38** [reply]
Kind of interesting how we had another user who had the same problem: https://github.com/mcguirepr89/BirdNET-Pi/issues/263, and his solution was also to do a complete reinstall.

From where I'm sitting it seems like purely a software issue, where birdnet.conf didn't get PRIVACY_THRESHOLD appended to it from update_birdnet_snippets.sh

Who knows. Best of luck.
